### ISSUE TITLE
SiriusAdapter with custom databases

### ISSUE THREAD

**Issue by heckendorfc** [OpenMS/OpenMS#4193]
Should be an easy fix but I'm not sure how you want to handle it.

`SiriusAdapter ... -sirius:database custom ...`
(output from Sirius via SiriusAdapter)
SEVERE14:22:17 - Unknown database 'custom'. Available are: training set, all, chebi, kegg, bio, ymdb, plantcyc, ecocyc mine, natural products, pubmed, hmdb, biocyc, kegg mine, hsdb, knapsack, biological, zinc bio, ymdb mine, gnps, unpd, pubchem, mesh, maconda, mydatabase

`SiriusAdapter ... -sirius:database mydatabase ...`
Invalid string parameter value 'mydatabase' for parameter 'database' given! Valid values are: 'all,chebi,custom,kegg,bio,natural products,pubmed,hmdb,biocyc,hsdb,knapsack,biological,zinc bio,gnps,pubchem,mesh,maconda'. Updating failed!

**Comment by oliveralka** [comment]
A custom database is currently not supported by the SIRIUSAdapter. 

In general and additional input would be possible, which then forces -sirius:database custom/mydatabase in the SIRIUS call. 

Do you know if that is usually used alone (custom only) or what happens if all is used in addition?

**Comment by heckendorfc** [comment]
I was testing with the latest OpenMS nightly and Sirius 4.0.1build9. I assume they just didn't update the documentation. It doesn't appear to be automatically generated from the list of available databases:
https://github.com/boecker-lab/sirius_frontend/blob/release/sirius_cli/src/main/java/de/unijena/bioinf/ms/cli/FingerIdOptions.java#L11

I've only used custom databases through the Sirius GUI, say for novel synthetic compounds, but I could imagine "all" being used if you want to check for those in addition to contaminants. I have not tried it.

I looked into generating a database through the CLI but didn't quickly find documented what input file formats Sirius is expecting, other than MOL which I didn't want to deal with. I ended up just using the GUI.

Would it make sense to just have -sirius:* (or some other notation? all args following `--`?) be passed as native Sirius args without validation by SiriusAdapter so you don't need to worry about keeping them in sync?

**Comment by timosachsenberg** [comment]
Hi,
While this sounds tempting, currently all possible command line parameters must be encoded in the tool itself. This allows us to guarantee that we can automatically generate nodes for several different workflow systems and automated documentation of parameters etc.
Allowing to directly pass the native arguments and skipping validation might sound like an option but I don't think that is a good idea. While parameters don't need to be kept in sync anymore, updating workflows, doing regression tests when new versions are released etc. will be much more difficult and produce harder to discover bugs if we support that as a first-class feature. That being said, I think we would still accept for certain tools like Sirius a pull request that provides that as an advanced option given.

**Comment by hroest** [comment]
I think however, that it could make sense as a hidden parameter for a special case like this. Not for routine useage but for special use cases where a user wants to test something specific?

**Comment by oliveralka** [comment]
As far as I understood this should be able to run with the SIRIUSAdapter e.g. you would get the mztab output and all of the data preprocessing. Just overwrite the additional parameters e.g. database for SIRIUS.

Another/additional approach could be to allow specifically named custom databases from the beginning e.g. custom_1, custom_2, ...., custom_10. These would have to be added to the path in SIRIUS (these have to be added anyway if they should be used by SIRIUS - as far as I know). If they are named in that manner it should be possible to use them. 

Unfortunately, I do not have any test data for such a case.
@heckendorfc do you by chance have a small dataset with a custom database you would be able to share for testing?

**Comment by heckendorfc** [comment]
That's an interesting idea, SiriusAdapter does currently work if I set the custom database name in Sirius to "custom". Numbered databases should work too but it would be difficult to keep track of what you have stored in each one. It might be possible to generate a list of available custom databases by looking through:
~/.sirius/csi_fingerid_cache/custom/

I'm still not sure what formats the Sirius CLI accepts but if you just paste your SMILES into the database generator in the GUI it will create it for you. Any compounds should work.
N#N
CN=C=O
[Cu+2].[O-]S(=O)(=O)[O-]

**Comment by oliveralka** [comment]
Yes, that is true. I guess this would work nicely for the CLI. The problem is that we ship OpenMS in different workflow systems here the parameters would need to be given beforehand and we are not able to generate them dynamically dependent on the path.
